EVENTS.md

```diff
@@ -0,0 +1,404 @@
+    sdk.NewAttribute("withdraw_address", withdrawAddr.String()),
+),
+```
+
+## Usage in wasmd
+
+In `x/wasm` we also use Events system. On one hand, the Go implementation of `x/wasm` emits standard events for each 
+message it processes, using the `distribution` module as an example. Furthermore, it allows contracts to
+emit custom events based on their execution state, so they can for example say "dex swap, BTC-ATOM, in 0.23, out 512"
+which require internal knowledge of the contract and is very useful for custom dApp UIs.
+
+`x/wasm` is also a consumer of events, since when a smart contract executes a SubMsg and processes the reply, it receives
+not only the `data` response from the message execution, but also the list of events. This makes it even more important for
+us to document a standard event processing format.
+
+### Standard Events in x/wasm
+
+Following the model of `distribution`, we will split the emitted events into two parts. All calls to the message server, will receive
+the following event:
+
+```go
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("module", "wasm"),
+    // Note: this was "signer" before 0.18
+    sdk.NewAttribute("sender", msg.Sender),  
+),
+```
+
+No further information will be added to the generic "message" type, but rather be contained in a more context-specific event type.
+Here are some examples:
+
+```go
+// Store Code
+sdk.NewEvent(
+    "store_code",
+    sdk.NewAttribute("code_id", fmt.Sprintf("%d", codeID)),
+    // features required by the contract (new in 0.18)
+    // see https://github.com/CosmWasm/wasmd/issues/574
+    sdk.NewAttribute("feature", "stargate"),
+    sdk.NewAttribute("feature", "staking"),
+)
+
+// Instantiate Contract
+sdk.NewEvent(
+    "instantiate",
+    sdk.NewAttribute("code_id", fmt.Sprintf("%d", msg.CodeID)),
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+// Execute Contract
+sdk.NewEvent(
+    "execute",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+// Migrate Contract
+sdk.NewEvent(
+    "migrate",
+    // Note: this is the new code id that is being migrated to
+    sdk.NewAttribute("code_id", fmt.Sprintf("%d", msg.CodeID)),
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+// Set new admin
+sdk.NewEvent(
+    "update_admin",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("admin", msg.NewAdmin),
+)
+
+// Clear admin
+sdk.NewEvent(
+    "clear_admin",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+)
+
+// Pin Code
+sdk.NewEvent(
+    "pin_code",
+    sdk.NewAttribute("code_id", strconv.FormatUint(msg.CodeID, 10)),
+)
+
+// Unpin Code
+sdk.NewEvent(
+    "unpin_code",
+    sdk.NewAttribute("code_id", strconv.FormatUint(msg.CodeID, 10)),
+)
+
+// Emitted when processing a submessage reply
+sdk.NewEvent(
+    "reply",
+    sdk.NewAttribute("_contract_addr", contractAddr.String()),
+    sdk.NewAttribute("success", strconv.FormatBool(err == nil)),
```

Comment:
Can you elaborate what "success" means in this context? When the reply handling fails with an error we revert the TX. 
To trace the replay call, we could store the submessage reply condition as an attribute: Always, Success, Error.

Reply:
This should be renamed to give more clarity.

When reply is called, it receives the result of a sub message processing, which is either "Ok" or "Err".

If I have RelyOn::Always, is is useful to see if I am processing a success call, or handling an error callback.
The main use-case is for devs having some visibility.

Do you have an idea for a better name?

Reply:
How about `("submsg", "ok")` and `("submsg", "err")` as the two options here? To show if we are processing a success case or handling an error.

Better suggestions welcome

Reply:
how about: "mode": "handle_success" || "handle_failure"